reactor-core/src/test/java/reactor/core/publisher/MonoUsingTest.java

```diff
@@ -36,6 +41,74 @@
+		return Arrays.asList(
+				Mono.using(() -> 1, Mono::just, cleanup::set, false),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.just(1), false)
+		);
+	}
+
+	public static List<Mono<Integer>> sourcesEager() {
+		return Arrays.asList(
+				Mono.using(() -> 1, Mono::just, cleanup::set),
+				Mono.using(() -> 1, Mono::just, cleanup::set, true),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.just(1)),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.just(1), true)
+		);
+	}
+
+	public static List<Mono<Integer>> sourcesFailNonEager() {
+		return Arrays.asList(
+				Mono.using(() -> 1, r -> Mono.error(new RuntimeException("forced failure")), cleanup::set, false),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.error(new RuntimeException("forced failure")), false)
+		);
+	}
+
+	public static List<Mono<Integer>> sourcesFailEager() {
+		return Arrays.asList(
+				Mono.using(() -> 1, r -> Mono.error(new RuntimeException("forced failure")), cleanup::set),
+				Mono.using(() -> 1, r -> Mono.error(new RuntimeException("forced failure")), cleanup::set, true),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.error(new RuntimeException("forced failure"))),
+				Mono.using(() -> cleanup::incrementAndGet, r -> Mono.error(new RuntimeException("forced failure")), true)
+		);
+	}
+
+	public static List<Mono<Object>> resourcesThrow() {
+		return Arrays.asList(
+				// non eager
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, Mono::just, cleanup::set, false),
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, r -> Mono.just(1), false),
+				// eager
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, Mono::just, cleanup::set),
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, Mono::just, cleanup::set, true),
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, r -> Mono.just(1)),
+				Mono.using(() -> { throw new RuntimeException("forced failure"); }, r -> Mono.just(1), true)
+		);
+	}
+
+	public static List<Mono<Integer>> sourcesThrowNonEager() {
+		return Arrays.asList(
+				Mono.using(() -> 1, r -> { throw new RuntimeException("forced failure"); }, cleanup::set, false),
+				Mono.using(() -> cleanup::incrementAndGet, r -> { throw new RuntimeException("forced failure"); }, false)
+		);
+	}
+
+	public static List<Mono<Integer>> sourcesThrowEager() {
+		return Arrays.asList(
+				Mono.using(() -> 1, r -> { throw new RuntimeException("forced failure"); }, cleanup::set),
+				Mono.using(() -> 1, r -> { throw new RuntimeException("forced failure"); }, cleanup::set, true),
+				Mono.using(() -> cleanup::incrementAndGet, r -> { throw new RuntimeException("forced failure"); }),
+				Mono.using(() -> cleanup::incrementAndGet, r -> { throw new RuntimeException("forced failure"); }, true)
+		);
+	}
+
+	private static final AtomicInteger cleanup = new AtomicInteger();
```

Comment:
Do you think we can avoid the static field? E.g. have the test cases contain the cleanup counter?

    static class CleanupCase {
        final Mono<Integer> resourceMono;
        final AtomicInteger cleanup;
        // ...
    }
    
And then have the test cases like

    public static List<CleanupCase> sourcesNonEager() {
        // ...
    }

I just want to prevent a situation where at some point we decide to use parallel tests and this would become a case to solve at that point.

Reply:
> we decide to use parallel tests and this would become a case to solve at that point.

oh it's really good point. ✅ updated by `CleanupCase`!